### PR TITLE
Implement climem memory monitor service

### PR DESCRIPTION
**Pull request by jasnell**
[`climem`](https://www.npmjs.com/package/climem) is a tool that I've used extensively in the past for diagnosing memory leaks in node.js applications. It provides simple monitoring of v8 memory usage and rss for the application and displays it in a chart in the terminal.

This PR implements support for climem directly within workerd. The workerd portion of it is implemented as a simple TCP server (per isolate) that, when connected, generates a json-serialized log of the current memory stats for the process/isolate. 

To use, specify the `-m localhost` option when running workerd serve. Then install `climem` from npm and run it, pointing at the tcp port that worker tells you the climem server is running on.

It's extremely rudimentary but that's the goal. Keep it simple.

/cc  @mcollina who wrote `climem`

### CODE REVIEW

**Comment by irvinebroque** [comment]
@admah — we should figure out the interface for Wrangler, so that you can say "turn on memory monitoring" and show this output in the terminal

File: src/workerd/util/mem_utils.c++ [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+#if !_WIN32
+kj::Maybe<size_t> tryGetResidentSetMemory() {
+  int fd_;
+  KJ_SYSCALL(fd_ = open("/proc/self/stat", O_RDONLY));
+  kj::AutoCloseFd fd(fd_);
+  auto text = kj::FdInputStream(kj::mv(fd)).readAllText();
+  auto p = text.begin();
+  for (kj::uint i = 0; i < 23; i++) {
+    p = strchr(p, ' ');
```
Comment (justin-mp) [inline src/workerd/util/mem_utils.c++:22]: Drive-by: `kj::String` has a `find` method now, though I don't know if it's any easier to use in this case.

File: src/workerd/io/worker.h [hunk 2]
```diff
@@ -334,6 +334,24 @@ class Worker::Isolate: public kj::AtomicRefcounted {
 
   kj::Own<const WeakIsolateRef> getWeakRef() const;
 
+  struct MemStats {
```
Comment (kentonv) [inline src/workerd/io/worker.h:337]: Why is this type declared separately in `Worker` and `jsg::IsolateBase`? Can both places share one type declaration?

In fact, could both just use `v8::HeapStatistics` directly instead of defining our own struct(s)? These stats are very specific to V8 anyway; it's not like a different JS engine is going to export the same statistics. So defining our own parallel structs seems to just add boilerplate IMO.

File: src/workerd/util/mem_utils.c++ [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+#if !_WIN32
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#endif
+
+namespace workerd::util {
+
+#if !_WIN32
```
Comment (kentonv) [inline src/workerd/util/mem_utils.c++:14]: This code is Linux-specific since it depends on procfs.

File: src/workerd/util/mem_utils.c++
Comment (kentonv) [inline src/workerd/util/mem_utils.c++:1]: Nitpick: We usually use hyphens rather than underscores in file names.

File: src/workerd/server/server.c++ [hunk 4]
```diff
@@ -1075,6 +1076,95 @@ private:
...
+                                     KJ_ASSERT_NONNULL(worker), name));
+        }
+
+        co_return;
+      }))();
+
+      kj::NEVER_DONE.wait(io.waitScope);
+    });
+    thread->detach();
+    threads.add(kj::mv(thread));
```
Comment (kentonv) [inline src/workerd/server/server.c++:1149]: Since you detached the thread, you can just discard the thread object.

**Comment by jasnell** [comment]
Meant to open this as a draft PR as it's not going to ready to land for a bit. Reviews are welcome but it's probably not going to be a super high priority to finish up.

File: src/workerd/util/mem_utils.c++ (hunk 3 above)
Comment (jasnell) [reply src/workerd/util/mem_utils.c++:14]: Yes, I know... this bit will need to be fully finished to work on macos/windows before the PR lands.

File: src/workerd/util/mem_utils.c++ (hunk 1 above)
Comment (jasnell) [reply src/workerd/util/mem_utils.c++]: Will look at that in a future iteration.
